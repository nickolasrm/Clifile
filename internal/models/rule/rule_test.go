package rule_test

import (
	"testing"

	"github.com/nickolasrm/clifile/internal/models/rule"
	. "github.com/nickolasrm/clifile/test/util"
	. "github.com/onsi/ginkgo/v2"
	. "github.com/onsi/gomega"
)

func TestRule(t *testing.T) {
	RegisterFailHandler(Fail)
	RunSpecs(t, "Rule Suite")
}

var _ = Describe("Rule", func() {
	When("a new rule is created", func() {
		var r *rule.Rule

		BeforeEach(func() {
			r = rule.NewRule("name", []string{"arg1", "arg2", "arg3"}, "doc", "action")
		})

		It("should return a variable type", func() {
			Expect(r).ToNot(BeNil())
		})
		When("Name() is called", func() {
			It("should return the rule name", func() {
				MatchSnapshot(r.Name())
			})
		})
		When("Positional() is called", func() {
			It("it should return the positional args", func() {
				MatchSnapshot(r.Positional())
			})
		})
		When("Doc() is called", func() {
			It("should return the rule doc", func() {
				MatchSnapshot(r.Doc())
			})
		})
		When("Actions() is called", func() {
			It("should return the rule actions", func() {
				MatchSnapshot(r.Actions())
			})
		})
		When("AppendActions() is called", func() {
			It("should append the text in actions as a newline", func() {
				r.AppendActions("newline\nnewline2")
				MatchSnapshot(r.Actions())
			})
		})
		When("AddRule() is called", func() {
			It("should add a new child rule", func() {
				r.AddRule(r)
				MatchSnapshot(r.Rules())
			})
		})
		When("Rules() is called", func() {
			When("no child rules exist", func() {
				It("should return an empty array", func() {
					MatchSnapshot(r.Rules())
				})
			})
			When("child rules exist", func() {
				It("should return the array of child rules", func() {
					r.AddRule(r)
					MatchSnapshot(r)
				})
			})
		})
		When("Rule() is called", func() {
			When("an existent child rule name is passed", func() {
				It("should return the rule", func() {
					r.AddRule(r)
					MatchSnapshot(r.Rule("name"))
				})
			})
			When("a non-existent child rule name is passed", func() {
				It("should return nil", func() {
					Expect(r.Rule("asd")).To(BeNil())
				})
			})
		})
	})
})
