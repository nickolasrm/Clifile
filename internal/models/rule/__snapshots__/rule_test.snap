
[Rule when a new rule is created when Name() is called should return the rule name - 1]
[]interface {}{
    "name",
}
---

[Rule when a new rule is created when Positional() is called it should return the positional args - 1]
[]interface {}{
    []string{"arg1", "arg2", "arg3"},
}
---

[Rule when a new rule is created when Doc() is called should return the rule doc - 1]
[]interface {}{
    "doc",
}
---

[Rule when a new rule is created when Actions() is called should return the rule actions - 1]
[]interface {}{
    "action\n",
}
---

[Rule when a new rule is created when AppendActions() is called should append the text in actions as a newline - 1]
[]interface {}{
    "action\nnewline\nnewline2\n",
}
---

[Rule when a new rule is created when AddRule() is called should add a new child rule - 1]
[]interface {}{
    []*rule.Rule{
        &rule.Rule{
            name:       "name",
            positional: {"arg1", "arg2", "arg3"},
            doc:        "doc",
            actions:    "action\n",
            rules:      {
                "name": &rule.Rule{(CYCLIC REFERENCE)},
            },
        },
    },
}
---

[Rule when a new rule is created when Rules() is called when no child rules exist should return an empty array - 1]
[]interface {}{
    []*rule.Rule{
    },
}
---

[Rule when a new rule is created when Rules() is called when child rules exist should return the array of child rules - 1]
[]interface {}{
    &rule.Rule{
        name:       "name",
        positional: {"arg1", "arg2", "arg3"},
        doc:        "doc",
        actions:    "action\n",
        rules:      {
            "name": &rule.Rule{(CYCLIC REFERENCE)},
        },
    },
}
---

[Rule when a new rule is created when Rule() is called when an existent child rule name is passed should return the rule - 1]
[]interface {}{
    &rule.Rule{
        name:       "name",
        positional: {"arg1", "arg2", "arg3"},
        doc:        "doc",
        actions:    "action\n",
        rules:      {
            "name": &rule.Rule{(CYCLIC REFERENCE)},
        },
    },
}
---
